src/vm/loaderallocator.cpp

```diff
@@ -501,10 +501,14 @@ LoaderAllocator * LoaderAllocator::GCLoaderAllocators_RemoveAssemblies(AppDomain
                 pAppDomain->RemoveFileFromCache(domainAssemblyToRemove->GetFile());
                 AssemblySpec spec;
                 spec.InitializeSpec(domainAssemblyToRemove->GetFile());
-                VERIFY(pAppDomain->RemoveAssemblyFromCache(domainAssemblyToRemove));
                 pAppDomain->RemoveNativeImageDependency(&spec);
             }
 
+            // We need to call RemoveAssemblyFromCache even for dynamic assemblies.
+            // Even though they don't have their own entries in the cache, they can
+            // be stored in the cache entries as parent assemblies.
+            VERIFY(pAppDomain->RemoveAssemblyFromCache(domainAssemblyToRemove));
```

Comment:
Could you write a test where what happens is 

1, From a collectible assembly A, you trigger a load of an assembly B in the default context
2. Unload the collectible assembly A, clearing this entry from the cache
3. Use Assembly.Load to load Assembly B again, and verify that you've loaded the same assembly B, and that there was still an entry in the AssemblySpecBindingCache for assembly B.

I worry that the scavenging process could cause us to delete entries from the cache which are relevant. The problem is that the parent assembly only is used for a relatively temporary amount of time, and its not unreasonable that due to the implementation of AssemblySpec::CopyFrom, that we might have just copied the parent assembly to assemblyspecs which are not interesting.

Reply:
Sure, I can do that after I get back from my vacation. I think you are right that it would very likely delete relevant entries. 
It seems that a fix would be to just to clear the parent assembly field in the cache entry's AssemblySpec field. Or is not having a parent assembly there an issue too and we would need to reparent such entry to something from the default context, e.g. the S.P.CoreLib?

Reply:
Well, if we're still in the process of loading the assembly, we can't safely clear the entry as we will read the value, but from my reading of the code we reach a point during load when we're done and we'll never read the field again (as long as we are in a successful load path.) Possibly that's when we could clear the parent assembly field. That leaves confusion for the case where the load was unsuccessful, where we attempt to re-generate an exception, etc. I'm not sure what to do there, possibly concluding on an unsuccessful load should have some effects. This is going to require careful testing, so I'd really like to see this submitted with a few specific test cases covering the scenarios identified.

Reply:
Thanks for all the insight, seems like another thing where I need to dive in again and get complete understanding of the subtle details.

Reply:
@davidwrighton I have re-read the code related to the AssemblySpec usage in the cache. It turns out that it is used only these ways:
* To locate entry in the cache, however the AssemblySpec::m_pParentAssembly is not used in the entry comparison.
* When a new entry is added to the cache (file, assembly or exception), the AssemblySpec is cloned and stored in the entry. The AssemblySpec that's in the entry is never accessed from the outside of the AssemblySpecBindingCache code.

So, based on this, we don't need to care about the parent assembly stored in the cache entry. To not to store possibly stale parent assembly in the entry, we can clear the parent assembly pointer after we clone the spec while storing it in the cache entry. It is not necessary as we don't access it anyways, but it seems it would be cleaner.

Then I can revert the added scavenging of cache entries based on the parent assembly.